### PR TITLE
Port file-locked-p

### PR DESCRIPTION
**Pull request by liff**
As a sort of continuation to #1394.

Some notes/caveats:
 - No Windows support. remacs doesn’t seem to build on Windows right now so there’s no way to test :frowning_face: 
 - Adds dependency to [systemstat crate](https://crates.io/crates/systemstat), which provides, among other things, system boot time information.
 - Some duplication between `filelock.c` and `filelock.rs` remain, pending ports of [`lock_file`](https://github.com/remacs/remacs/blob/fa68c1f5df7834d063e8b9c7b8a262178e0059ca/src/filelock.c#L640), [`unlock_file`](https://github.com/remacs/remacs/blob/fa68c1f5df7834d063e8b9c7b8a262178e0059ca/src/filelock.c#L725), [`unlock_all_files`](https://github.com/remacs/remacs/blob/fa68c1f5df7834d063e8b9c7b8a262178e0059ca/src/filelock.c#L742) and [`unlock_buffer`](https://github.com/remacs/remacs/blob/fa68c1f5df7834d063e8b9c7b8a262178e0059ca/src/filelock.c#L757).

### CODE REVIEW

**Comment by brotzeit** [comment]
Wow =)

File: rust_src/src/filelock.rs [hunk 1]
```diff
@@ -2,13 +2,277 @@
...
+        // See https://bugzilla.redhat.com/show_bug.cgi?id=1384153
+        let boot_time_sep = |c| c == ':' || c == '\u{F022}';
+
+        // The USER is everything before the last @.
+        let (user, rest) = data.split_at(data.rfind('@')?);
+        let user = user.to_string();
+
+        // The HOST is everything after the '@' to the last '.'.
+        let (at_host, rest) = rest.split_at(rest.rfind('.')?);
+        let host = at_host[1..].to_string();
```
Comment (shaleh) [inline rust_src/src/filelock.rs]: Why the slicing of `at_host` when `user` above could be used directly?

File: rust_src/src/filelock.rs [hunk 2]
```diff
@@ -2,13 +2,277 @@
...
+    /// The file is locked by this Emacs process.
+    LockedByUs,
+    /// The file is locked by some other Emacs process identified
+    /// by [`LockInfo`].
+    LockedBy(LockInfo),
+}
+
+/// Opens a file in `O_NOFOLLOW` mode. Returns [`None`] if the open failed
+/// because the `path` refers to a symbolic link.
+#[cfg(unix)]
```
Comment (shaleh) [inline rust_src/src/filelock.rs]: Can you add stubs for windows please? That when someone wants Windows to work we have all of the steps laid out for them.

File: rust_src/src/filelock.rs [hunk 3]
```diff
@@ -2,13 +2,277 @@
...
+fn open_nofollow(path: &Path) -> Result<Option<File>> {
```
Comment (shaleh) [inline rust_src/src/filelock.rs]: Instead of including here could you add this to the top with a `cfg(unix)` wrapper?

File: rust_src/src/filelock.rs [hunk 4]
```diff
@@ -2,13 +2,277 @@
...
+        fd => {
+            let file = unsafe { File::from_raw_fd(fd) };
+            Ok(Some(file))
+        }
+    }
+}
+
+/// Reads the contents of a regular file at `path` up to `max_content_length` bytes.
+/// Returns [`None`] if the `path` is a symbolic link.
+fn read_nofollow(path: &Path, max_content_length: usize) -> Result<Option<String>> {
```
Comment (shaleh) [inline rust_src/src/filelock.rs]: I am on the fence. I think saving the result into a variable and returning `Ok(result)` makes it more clear that the only failures here are from the reading and parsing. 

```
let result = match open_nofollow(path)? {
    Some(file) => {
        let mut content = String::with_capacity(max_content_length);
            file.take(max_content_length as u64)
                .read_to_string(&mut content)?;
        Some(content)
    }
    None => None,
};
Ok(result)
```

File: rust_src/src/filelock.rs [hunk 5]
```diff
@@ -2,13 +2,277 @@
...
+        }
+
+        Err(ref e) if e.kind() == NotFound => Ok(None),
+
+        Err(e) => Err(e),
+    }
+}
+
+/// Returns [`true`] if a process with PID `pid` is running.
+#[cfg(unix)]
```
Comment (shaleh) [inline rust_src/src/filelock.rs]: Windows stub here too.

File: rust_src/src/filelock.rs [hunk 6]
```diff
@@ -2,13 +2,277 @@
...
+        0 => true,
+        // `PermissionDenied` indicates that there _is_ a process but it’s
+        // owned by another user
+        _ => Error::last_os_error().kind() == PermissionDenied,
+    }
+}
+
+/// Returns the system’s last boot time as Unix epoch seconds.
+fn get_boot_time() -> Result<i64> {
+    use systemstat::Platform;
```
Comment (shaleh) [inline rust_src/src/filelock.rs]: Why is this import here and not at the top?

File: rust_src/src/filelock.rs [hunk 7]
```diff
@@ -2,13 +2,277 @@
...
+    let sys = systemstat::System::new();
+    sys.boot_time().map(|t| t.timestamp())
+}
+
+/// Returns [`true`] if there is no boot time in the lock info or if the
+/// boot time is within one second of the system’s boot time.
+fn boot_time_within_one_second(info: &LockInfo) -> bool {
```
Comment (shaleh) [inline rust_src/src/filelock.rs]: Great naming and easy to follow logic.

File: rust_src/src/filelock.rs [hunk 8]
```diff
@@ -2,13 +2,277 @@
...
+    info.boot_time.map_or(true, |boot_time| {
+        let system_boot_time = get_boot_time().unwrap_or(0);
+        (boot_time - system_boot_time).abs() <= 1
+    })
+}
+
+/// Returns the current state of the lock from lock file `path`.
+fn current_lock_owner(path: &Path) -> Result<LockState> {
+    use LockState::*;
```
Comment (shaleh) [inline rust_src/src/filelock.rs]: Why the local import?

File: rust_src/src/filelock.rs [hunk 9]
```diff
@@ -2,13 +2,277 @@
...
+
+    match read_lock_info(path)? {
```
Comment (shaleh) [inline rust_src/src/filelock.rs]: This is another example where perhaps saving the result and returning `Ok(result)` is easier to follow.

File: rust_src/src/filelock.rs [hunk 10]
```diff
@@ -43,4 +307,112 @@ pub fn unlock_buffer_lisp() {
...
+
+    match current_lock_owner(&path) {
+        Ok(NotLocked) | Err(_) => Qnil,
+        Ok(LockedByUs) => Qt,
+        Ok(LockedBy(info)) => LispObject::from(info.user.as_str()),
+    }
+}
+
 include!(concat!(env!("OUT_DIR"), "/filelock_exports.rs"));
+
```
Comment (shaleh) [inline rust_src/src/filelock.rs]: So nice to see some native test cases.

File: rust_src/src/filelock.rs [hunk 11]
```diff
@@ -2,13 +2,277 @@
...
+        // The HOST is everything after the '@' to the last '.'.
+        let (at_host, rest) = rest.split_at(rest.rfind('.')?);
+        let host = at_host[1..].to_string();
+
+        // The PID is everything from the last '.' to the ':' or equivalent.
+        // NOTE: Unlike GNU Emacs, we do not tolerate negative PID values
+        // or over/underflow.
+        let (pid_str, boot_time_str) =
+            rest.split_at(rest.find(boot_time_sep).unwrap_or(rest.len()));
+        let pid = pid_str[1..].parse::<i32>().ok().filter(|pid| pid > &0)?;
```
[the remark this answers is not in the document]
Comment (liff) [reply rust_src/src/filelock.rs]: [`parse()`](https://doc.rust-lang.org/std/primitive.str.html#method.parse) returns a `Result` and [`ok()`](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) converts it to an `Option` so we just lose the error details, not panic.

**Comment by liff** [comment]
Thanks for the review. Updated per your comments.

**Comment by brotzeit** [comment]
@liff You can merge your PR when you think it's ready.
